src/account/operations/participation/mod.rs

```diff
@@ -63,55 +64,70 @@ impl AccountHandle {
             })
             .collect::<Vec<OutputData>>();
 
+        let mut events = HashMap::new();
+        for output_data in participation_outputs {
+            // PANIC: the filter already checks that the metadata exists.
+            let metadata = output_data.output.features().and_then(|f| f.metadata()).unwrap();
+
+            let event_ids = if let Ok(participations) = Participations::from_bytes(&mut metadata.data()) {
+                participations
+                    .participations
+                    .into_iter()
+                    .map(|p| p.event_id)
+                    .collect::<Vec<ParticipationEventId>>()
+            } else {
+                // No valid participation in this output, skip it.
+                continue;
+            };
+
+            for event_id in event_ids {
+                match events.entry(event_id) {
```

Comment:
Maybe put that in a loop over the participations so you don't need the collect L77?